Add setting bpf per outzq in zbalance_ipc

EaseTheWorld opened the pull request:
This demonstrates pfring_zc_set_bpf_filter for inzq and outzq.
  -f bpf1 : setting bpf file 'bpf1' for all inzq
  -f bpf1@0,1 -f bpf2@2 : setting bpf file bpf1 for outzq 0,1 and bpf2 for outzq 2(other outzq is not affected)
This uses idle_func and filter_func so I always set idle_func and filter_func for simplicity.

Now zbalance_ipc becomes a software NPB(Network Packet Broker) now.
But remember bpf is expensive. wire-speed is not guaranteed with bpf.

EaseTheWorld committed:
Add setting bpf per outzq in zbalance_ipc

This demonstrates pfring_zc_set_bpf_filter for inzq and outzq.
  -f bpf1 : setting bpf file 'bpf1' for all inzq
  -f bpf1@0,1 -f bpf2@2 : setting bpf file bpf1 for outzq 0,1 and bpf2 for outzq 2(other outzq is not affected)
This uses idle_func and filter_func so I always set idle_func and filter_func for simplicity.

Now zbalance_ipc becomes a software NPB(Network Packet Broker) now.
But remember bpf is expensive. wire-speed is not guaranteed with bpf.

cardigliano commented:
Thank you @EaseTheWorld, nice feature, a couple of comments:
- I would set filter_func only when some filter has been configured (including bpf) to avoid affecting the performance when no filter is used
- please use meaningful names (e.g. it is not clear that readline_from_file_list is related to bpf and what it does)

EaseTheWorld commented:
- if bpf is not configured, filter_func is just simply return because inzq_bpf or outzq_bpf is NULL.
  simply setting filter_func(which does nothing) affects performance internally?
- name was also intentional to show it is just static utility(fileIO/linked list) function. but I'll change it.

cardigliano commented:
Setting filter_func to NULL avoids calling the function and evaluating a few "if" conditions, this is important at high rates..

EaseTheWorld committed:
Add setting bpf per outzq in zbalance_ipc 2

- rename bpf-related structure and functions more readable
- setting idle_func, filter_func, signal handler only if bpf is specified.

cardigliano commented:
Thank you
